### DISCUSSION TITLE
Why radius of ewa_* 4tap is 4.241?

### DISCUSSION THREAD

**Discussion by rainlycoris** [mpv-player/mpv#14325] [Q&A] [answered]
How did get this value? I know 3taps is 3.28..., 4taps is 4.2410628637960698819573.
And why many decimal places at blur of `ewa_lanczossharp` or `ewa_lanczos4sharpest`? I believe this is value isn't made by human eyes.
What is the effect if I don't apply blur, Like `ewa_lanczos4`? That is also what I'm curious about. Why isn't have this algorithm in mpv.

**Comment by llyyr** [comment]
https://code.videolan.org/videolan/libplacebo/-/merge_requests/560#note_408215

**Reply by rainlycoris** [reply]
Thank you very much! And answer of first question is this value in `ewa_*` can achieve normal 4taps right? Would you briefly explain the principle? Thanks.

**Reply by rainlycoris** [reply]
And this code run 3taps obtain `0.88451209326050`, dosen't equal `0.98...` at `ewa_lanczossharp `.

**Reply by rainlycoris** [reply]
> And this code run 3taps obtain `0.88451209326050`, dosen't equal `0.98...` at `ewa_lanczossharp `.

I got it, this is why `sharpest` different with `sharp`.

**Comment by rainlycoris** [comment]
Say some digressions, If I used ewa_hanning with 4taps, did I make a mistake in math?
My goal is to make it clearer.
It has to be said that this looks like still less ringing.

**Reply by rainlycoris** [reply]
This is my fantasy, there is no observable difference between 3tap and 4tap, even at the pixel level.

my screenshot was wrong, I discovered that the built-in screenshot of the mpv does not include zooming, the correct screenshot below.
<img width="1512" alt="截屏2024-06-16 17 43 51" src="https://github.com/mpv-player/mpv/assets/125644853/5efce36c-8052-4fe9-ab01-676e3de4f281">
ewa_hanning 3tap

<img width="1512" alt="截屏2024-06-16 17 44 33" src="https://github.com/mpv-player/mpv/assets/125644853/c7fd6df8-de38-430a-a1a8-bc4ac813c023">
ewa_hanning 4tap

<img width="1512" alt="截屏2024-06-16 17 45 36" src="https://github.com/mpv-player/mpv/assets/125644853/e7eaafb4-af0b-41b3-8e85-9cf15a0abd16">
ewa_lanzcos

**Reply by kasper93** [reply]
Can you post exact commands you used? And what is the goal you want to achieve?

**Reply by rainlycoris** [reply]
> Can you post exact commands you used? And what is the goal you want to achieve?

My goal is to make it sharper, And control ringing problem such as used ewa_hanning.

My current problem is whether ewa_hanning 4tap is sharper than 3tap. unfortunately, I didn't see it. Is this my illusion? because 4tap ewa_lanzcos boost significantly than 3tap. 4tap ewa_hanning should be better on math?

I used
```
scale = ewa_hanning
scale-radius = 3.2383154841662362076499/4.2410628637960698819573 (For the example I gave)
cscale = catmull_rom
dither-depth = 10
correct-downscaling = yes
linear-downscaling = yes
sigmoid-upscaling = yes
deband = yes
```

I have two more questions for you, by the way. If I used `scaler-lut-size = 10`（Although I knew now this selection is hard-coded, and value is 8）and `fbo-format = rgba32f` can be better?

Thank you very much!

**Reply by ghost** [reply, marked as answer]
I'm going to be completely honest, there's no benefit to using 4-lobe `ewa_hanning` over 3-lobe `ewa_lanczossharp`. The sharpness difference isn't perceivable but the former rings very badly due to having a more prominent halo. This kind of defeats the purpose of even adding the Hann window in the first place.

There isn't even that much of a benefit to using 4-lobe `ewa_lanczos`, hence why `ewa_lanczos4` will never exist as an aliased option in mpv. The only reason why `ewa_lanczos4sharpest` even has a usecase is because the 3-lobe "sharpest" variant of the filter is subpar for upscaling and has nasty aliasing for a polar filter. Meanwhile the 4-lobe variant performs better in comparison (though it isn't strictly better than `ewa_lanczossharp`, it just has different usecases).

What you're actually looking for is either antiringing and/or a sharpening filter, as you'll just get diminishing returns by trying to tweak built-in scalers. Many people have already played around with this for decades (some even made careers revolving around this) with very minimal success. At a certain point you just have to throw in the towel and accept that `lanczos` is endgame. :)

**Reply by rainlycoris** [reply]
I think you're absolutely right.

So what should I use? `ewa_hanning` or `ewa_lanczoss with anti-ringing` or just `lanczos` (`ewa_lanczos` is strictly better that `lanczos` right?). But [mpv handbook](https://github.com/mpv-player/mpv/blob/master/DOCS/man/options.rst) said that anit-ringing isn't affect any polar scalers. I'm surprised that 4sharpest does exactly that and used exactly this parameter.

Is there anything else I need to do on scale, such as for chroma? No one seems to have come up with a best chroma or dscale routine algorithm . Of course, in fact, I can't see much difference, I just want to get the best result in math. I'm sure there are others who feel the same way, especially now that computer power is quite adequate. thank you for your help.

**Reply by ghost** [reply]
> So what should I use?

`ewa_lanczossharp` is generally accepted to be a good all-rounder filter, and combining it with antiringing is popular among users.
> ewa_lanczos is strictly better that lanczos right?

It has less prominent ringing and less aliasing in exchange for being slightly blurrier than tensor `lanczos`. Both filters have acceptable uses but `ewa_lanczossharp` is generally higher quality for most people.

From sharper (more ringing) to blurrier (less ringing):
`ewa_lanczos4sharpest` > `lanczos` > `ewa_lanczossharp` > `ewa_hanning`

Anti-aliasing (this ranking is the same for hash pattern dampening):
`ewa_lanczossharp` > `ewa_hanning` > `ewa_lanczos4sharpest` > `lanczos`

> But [mpv handbook](https://github.com/mpv-player/mpv/blob/master/DOCS/man/options.rst) said that anit-ringing isn't affect any polar scalers.

Antiringing only works with polar filters when using `--vo=gpu-next`. `--scale-antiring=0.8` is hardcoded for `ewa_lanczos4sharpest` because most people agree that it looks bad without antiringing.

> No one seems to have come up with a best chroma or dscale routine algorithm

If performance isn't an issue then there's no point in using different filters for `--scale` and `--cscale`. 

mpv's default `--dscale=hermite` is already "high quality" but there isn't a perfect solution for downscaling, as downscaling is a balancing act between sharpness and aliasing. The only issue with `hermite` is that it can exhibit moiré patterns at high downscale factors due to its small radius, but this isn't an issue for most people. You can play around with some polar filters, but these are exceptionally slow when downscaling, and some of the better polar filters require antiringing when scaling in linear light (which just makes the filter even slower).

**Reply by rainlycoris** [reply]
I’ve found that `ewa_hanning4` seems like a barely acceptable option. Unlike the `Jinc` window, the `Hann` window has a useful property: its relative sidelobe level is essentially independent of the tap count; increasing the number of taps mainly narrows the main lobe. In terms 
of sharpness, `ewa_lanczos4` > `ewa_hanning4` > `ewa_lanczos` > `ewa_hanning`,
of ringing, `ewa_hanning` < `ewa_lanczos` < `ewa_hanning4` < `ewa_lanczos4`,
of aliasing, `ewa_hanning` < `ewa_hanning4` < `ewa_lanczos` < `ewa_lanczos4`.
Unfortunately, `ewa_hanning4` may be inferior to `ewa_lanczossharp`. I believe [ghost](https://github.com/ghost) is entirely right on this.

I'm currently using `ewa_hanning3` with `scale-blur=0.9753886283057314`, calculated to target 
a Nyquist frequency of 0.5.
It yields rather pleasing results: ringing ≈ `ewa_hanning`, sharpness ≈ `ewa_lanczossharp`. [Research Report.md](https://github.com/user-attachments/files/31294059/ewa_hanning_blur_research_report.md) 
[Research Report (English Ver).md](https://github.com/user-attachments/files/31325015/ewa_hanning_blur_research_report_en.md).


<img width="2000" height="1906" alt="1" src="https://github.com/user-attachments/assets/621344d6-e57d-4dd4-8a05-109ffb26a4b0" />

**Reply by kkanungo17** [reply]
Are you using scale-antiring with your new parameters?

**Reply by rainlycoris** [reply]
> Are you using scale-antiring with your new parameters?

No. I'm using the `vo=libmpv` backend, and it wouldn't work there anyway.

**Reply by kkanungo17** [reply]
I see, then it's not using the libplacebo filters.

**Reply by rainlycoris** [reply]
Sorry to bother you. Could I get your thoughts on these parameters and the algorithm's objective? My original goal was the same as `ewa_lanczossharp`, to make L1 no-op / `No-Op Vertical and Horizontal Line Preservation Condition`, but on `ewa_hanning` the Nyquist > 0.5.
It looks fine to me visually, though human eyes aren't that sensitive. I'm not entirely sure if I picked the best quantitative metrics, but the numbers look good on paper. What are your thoughts?

**Reply by HedioKojima** [reply]
Either ewa_hanning or ewa_lanczos can be used; personal preferences may vary, and of course mathematically one is indeed better than the other. But is it really necessary to use them for scaling? In most cases, FSR or RAVU_lite should be sufficient for general needs.
